File: phalcon/service/Socket.php
<?php
namespace Service;

use Config\Env;
use Service\ApiToken;
use Service\AdminToken;
use Util\Util;

use Ratchet\WebSocket\MessageComponentInterface;
use Ratchet\ConnectionInterface;

/* 数据类 */
class Socket implements MessageComponentInterface {

  public $clients=null; //连接
  public $uids=[];      //Uid

  /* 构造函数 */
  function __construct() {
    $this->clients = new \SplObjectStorage;
  }

  /* 路由 */
  function router(string $uid, $conn, $msg): void {
    $data = json_decode($msg, true);
    if($data['type']=='msg'){
      $this->getMsg($uid, $conn, $data);
    }else{
      $this->send($conn, ['code'=>0, 'type'=>'', 'msg'=>'成功']);
    }
  }

  /* 消息 */
  function getMsg(string $uid, $conn, $msg) {
    print_r($uid);
    print_r($msg);
    // 群发
    if($uid=='0') {
      $this->sendAll($msg);
      return;
    }
    // 单发
    $this->send($conn, [
      'code'=> 0,
      'type'=> 'msg',
      'msg'=> '成功',
      'time'=> date('Y-m-d H:i:s'),
    ]);
  }

  /* 群发 */
  function sendAll($data) {
    foreach ($this->clients as $conn) {
      $this->send($conn, $data);
    }
  }

  /* 单发 */
  function send($conn, $data) {
    $res = json_encode($data);
    $conn->send($res);
  }

  /* 连接 */
  function onOpen(ConnectionInterface $conn) {
    // 验证
    $uid = $this->verify($conn);
    if($uid==''){
      return $conn->close();
    }
    // 连接
    $this->clients->attach($conn);
    $this->uids[$uid] = $conn->resourceId;
  }

  /* 验证 */
  private function verify($conn): string {
    // 参数
    $param = (array)$conn->httpRequest->getUri();
    $data = [];
    foreach($param as $val) $data[] = $val;
    $arr = Util::UrlToArray($data[5]);
    if(empty($arr)) return '';
    $type = $arr['type'];
    $token = $arr['token'];
    if(empty($type) || empty($token)) return '';
    // 验证
    if($token==Env::$key){
      return '0';
    }elseif($type=='api'){
      $tData = ApiToken::token($token);
      if(empty($tData)) return '';
      return (string)$tData->uid;
    }elseif($type=='admin'){
      $tData = AdminToken::token($token);
      if(empty($tData)) return '';
      return (string)$tData->uid;
    }
    return '';
  }

  /* 关闭 */
  function onClose(ConnectionInterface $conn) {
    $this->clients->detach($conn);
    // 移除Uid
    foreach($this->uids as $key=>$val) {
      if($val==$conn->resourceId){
        unset($this->uids[$key]);
        break;
      }
    }
  }
  /* 错误 */
  function onError(ConnectionInterface $conn, \Exception $e) {
    return $conn->close();
  }

  /* 消息 */
  function onMessage(ConnectionInterface $from, $msg) {
    // 验证
    $uid = $this->verify($from);
    if($uid==''){
      return $from->close();
    }
    // 路由
    $this->router($uid, $from, $msg);
  }

}

File: java/src/main/java/webmis/service/Socket.java
package webmis.service;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;

import javax.websocket.OnClose;
import javax.websocket.OnMessage;
import javax.websocket.OnOpen;
import javax.websocket.Session;
import javax.websocket.server.ServerEndpoint;

import com.alibaba.fastjson.JSONObject;

import org.springframework.stereotype.Controller;

import webmis.config.Env;
import webmis.util.Util;

@Controller
@ServerEndpoint("/websocket")
public class Socket extends Base {

  public static ConcurrentHashMap<String, Session> clients = new ConcurrentHashMap<>(); //连接

  /* 路由 */
  @SuppressWarnings("unchecked")
  public void router(String uid, Session conn, String msg) {
    HashMap<String, Object> data = JSONObject.parseObject(msg, HashMap.class);
    if(data.get("type").equals("msg")){
      getMsg(uid, conn, data);
    }else{
      HashMap<String, Object> res = new HashMap<String, Object>();
      res.put("type", "");
      res.put("code", 0);
      res.put("msg", "成功");
      send(conn, res);
    }
  }

  /* 消息 */
  public void getMsg(String uid, Session conn, HashMap<String, Object> msg) {
    Print(uid, msg);
    // 群发
    if(uid.equals("0")) {
      sendAll(msg);
      return ;
    }
    // 单发
    HashMap<String, Object> data = new HashMap<String, Object>();
    data.put("code", 0);
    data.put("type", "msg");
    data.put("msg", "成功");
    data.put("time", Util.Date(""));
    send(conn, data);
  }

  /* 群发 */
  public void sendAll(HashMap<String, Object> data) {
    for(Entry<String, Session> obj : clients.entrySet()) {
      send(obj.getValue(), data);
    }
  }

  /* 单发 */
  public void send(Session conn, HashMap<String, Object> data) {
    String res = Util.JsonEncode(data);
    try {
      conn.getBasicRemote().sendText(res);
    } catch (IOException e) {
      Print("[Socket] Send:", e.getMessage());
    }
  }

  /* 链接 */
  @OnOpen
  public void onOpen(Session session) {
    // 验证
    String uid = verify(session.getRequestParameterMap());
    // 保存
    clients.put(uid, session);
  }
  /* 消息 */
  @OnMessage
  public void onMessage(String message, Session session) {
    // 验证
    String uid = verify(session.getRequestParameterMap());
    if(uid==""){
      try { session.close(); } catch (IOException e) {}
      return;
    }
    // 路由
    router(uid, session, message);
  }
  /* 关闭 */
  @OnClose
  public void onClose(Session session) {
    String uid;
    Session conn;
    // 移除
    for(Entry<String, Session> obj : clients.entrySet()) {
      uid = obj.getKey();
      conn = obj.getValue();
      if(conn == session){
        clients.remove(uid);
        break;
      }
    }
  }

  /* 验证 */
  private String verify(Map<String, List<String>> param) {
    // 参数
    String type = param.containsKey("type")?param.get("type").get(0):"";
    String token = param.containsKey("token")?param.get("token").get(0):"";
    if(type.isEmpty() || token.isEmpty()) return "";
    // 验证
    if(token.equals(Env.key)){
      return "0";
    }else if(type.equals("api")){
      HashMap<String, Object> tData = ApiToken.Token(token);
      if(tData.isEmpty()) return "";
      return tData.get("uid").toString();
    }else if(type.equals("admin")){
      HashMap<String, Object> tData = AdminToken.Token(token);
      if(tData.isEmpty()) return "";
      return tData.get("uid").toString();
    }
    return "";
  }
  
}
